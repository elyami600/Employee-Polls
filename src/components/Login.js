import { useState } from "react";
import { connect } from "react-redux";
import { setAuthedUser } from "../actions/authedUser";
import { Navigate, useNavigate, useLocation } from "react-router-dom";




const LoginPage = ({ dispatch, users, loggedIn }) => {
    const navigate = useNavigate();
    const { state } = useLocation();

    const [error, setError] = useState(false);
    const [username, setUsername] = useState('');
    const [password, setPassword] = useState('');
    const [isOpen, setIsOpen] = useState(false);
    
    

    // Redirect if logged in
    if (loggedIn) {
        const params = new URLSearchParams(window.location.search);
        const redirectURL = params.get('redirectTo');
        return <Navigate to={redirectURL || "/"} />;
    }
    // if (loggedIn) {
    //     const lastVisitedURL = localStorage.getItem('lastVisitedURL');
    //     return <Navigate to={lastVisitedURL || "/"} />;
    // }


    // Single handler for both inputs
    const handleInputChange = (e) => {
        const { name, value } = e.target;
        if (name === "username") setUsername(value);
        else if (name === "password") setPassword(value);
    };
   // handler the submit  for both inputs
   // ckeck if the user exits
    const handleSubmit = (e) => {
        e.preventDefault();
        const userExists = users.find(
            (user) => user.id === username && user.password === password
        );

        if (userExists) {
            dispatch(setAuthedUser(username));
             navigate(state?.path || "/");
            // const redirectTo = state?.path || localStorage.getItem('lastVisitedURL') || "/";
            // navigate(redirectTo);
            //localStorage.removeItem('lastVisitedURL');
        } else {
            setError(true);
            setUsername("");
            setPassword("");
        }
    };

    const toggleDropdown = () => {
        setIsOpen(!isOpen);
      };

      
      const handleUserSelect = (user) => {
        setUsername(user.id);
        setPassword(user.password);
        setIsOpen(false); 
        setError(false); 
    };
    

    return (
        <div className="container">
            {error && (
                <h1 className="center" data-testid="error-header">
                    Error: Invalid username or password.
                </h1>
            )}
            
            <h1 className="center">Log In</h1>
            <button className="existUser-btn" onClick={toggleDropdown}><strong>Exist User</strong></button>
           
            <form className="center" onSubmit={handleSubmit}>
                <div>
                    <label>Username:</label>
                    <input 
                        data-testid="username-input"
                        type="text"
                        placeholder="username..."
                        name="username"
                        value={username}
                        onChange={handleInputChange}
                    />
                </div>
                <br/>
                <div>
                    <label>Password:</label>
                    <input 
                        data-testid="password-input"
                        type="password"
                        placeholder="password..."
                        name="password"
                        value={password}
                        onChange={handleInputChange}
                    />
                </div>
                <br/>
                <br/>
                {isOpen && (
                <ul className="center">
                    {users.map((user) => (
                        <li key={user.id} onClick={() => handleUserSelect(user)}>
                            {user.name}
                        </li>
                    ))}
                </ul>
            )}  <br/>
               <br/>
                <button
                    data-testid="submit-button"
                    className="btn"
                    type="submit"
                    disabled={!username || !password}
                >
                    Submit
                </button>

            </form>
    
           
            
        </div>
    );
};

const mapStateToProps = ({ users, authedUser }) => ({
    users: Object.values(users),
    loggedIn: !!authedUser,
});

export default connect(mapStateToProps)(LoginPage);
