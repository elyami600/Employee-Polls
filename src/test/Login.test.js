import * as React from 'react';
import { render, screen, fireEvent, userEvent} from '@testing-library/react';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import reducers from '../reducers'; 
import middleware from "../middleware"
import Login from '../components/Login';
import { MemoryRouter } from "react-router";

const store = createStore(reducers, middleware);

describe('Login' , () => {
    const component = () => {
        render(
            <MemoryRouter>
                <Provider store={store}>
                    <Login />
                </Provider>
            </MemoryRouter>
        );
    };
 
    it("renders the component", () => {
        const { container } = render(
            <MemoryRouter>
                <Provider store={store}>
                    <Login />
                </Provider>
            </MemoryRouter>
        );
        expect(container).toBeDefined();
        expect(container).toMatchSnapshot();
    });

    it("renders username, password inputs, and submit button", () => {
        component()

        expect(screen.getByTestId('username-input')).toBeInTheDocument();
        expect(screen.getByTestId('password-input')).toBeInTheDocument();
        expect(screen.getByTestId('submit-button')).toBeInTheDocument();
        expect(screen.queryByTestId('error-header')).not.toBeInTheDocument();
    });

    it("submits form with valid username and password", () => {
        component()
       
        const username = screen.getByTestId('username-input')
        const password = screen.getByTestId('password-input')
        const button   = screen.getByTestId('submit-button')

        fireEvent.change(username, 'mtsamis');
        fireEvent.change(password, 'xyz123');
        fireEvent.click(button, { name: "Submit" });

        expect(screen.queryByTestId('error-header')).not.toBeInTheDocument();
    }); 

    it("displays an error message if the username and password do not match", () => {
        component()

        const username = screen.getByTestId('username-input')
        const password = screen.getByTestId('password-input')
        const button   = screen.getByTestId('submit-button')

        fireEvent.change(username, { target: { value: 'tylermcginnis' } });
        fireEvent.change(password, { target: { value: 'incore' } });
        fireEvent.click(button);

        expect(screen.queryByTestId('error-header')).toBeInTheDocument();
    });

    it("disables submit button if username or password fields are empty", () => {
       component()
       const username = screen.getByTestId('username-input')
       const password = screen.getByTestId('password-input')
       const button   = screen.getByTestId('submit-button')

       fireEvent.change(username, { target: { value: '' } });
       fireEvent.change(password, { target: { value: '' } });

       expect(button).toBeDisabled();

   });
    
})





