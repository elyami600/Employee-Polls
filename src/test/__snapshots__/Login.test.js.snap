// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders the component 1`] = `
<div>
  <div
    class="container"
  >
    <h1
      class="center"
    >
      Log In
    </h1>
    <button
      class="existUser-btn"
    >
      <strong>
        Exist User
      </strong>
    </button>
    <form
      class="center"
    >
      <div>
        <label>
          Username:
        </label>
        <input
          data-testid="username-input"
          name="username"
          placeholder="username..."
          type="text"
          value=""
        />
      </div>
      <br />
      <div>
        <label>
          Password:
        </label>
        <input
          data-testid="password-input"
          name="password"
          placeholder="password..."
          type="password"
          value=""
        />
      </div>
      <br />
      <br />
        
      <br />
      <br />
      <button
        class="btn"
        data-testid="submit-button"
        disabled=""
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</div>
`;
